src/Control/Middleware/RateLimitMiddleware.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+namespace SilverStripe\Control\Middleware;
+
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\HTTPResponse;
+use SilverStripe\Core\Cache\RateLimiter;
+use SilverStripe\ErrorPage\ErrorPage;
+use SilverStripe\ORM\FieldType\DBDatetime;
+use SilverStripe\Security\Security;
+
+class RateLimitMiddleware implements HTTPMiddleware
+{
+
+    /**
+     * @param HTTPRequest $request
+     * @param callable $delegate
+     * @return HTTPResponse
+     */
+    public function process(HTTPRequest $request, callable $delegate)
+    {
+        $limiter = new RateLimiter($this->getKeyFromRequest($request), 10, 1);
+        if ($limiter->canAccess()) {
+            $limiter->hit();
+            $response = $delegate($request);
+        } else {
+            $response = $this->getErrorHTTPResponse();
+        }
+        $this->addHeadersToResponse($response, $limiter);
+        return $response;
+    }
+
+    /**
+     * @param HTTPRequest $request
+     * @return string
+     */
+    protected function getKeyFromRequest($request)
```

Comment:
The rate limiting is applied to a particular route, but seems to be global in it's effect. AFAICT, it doesn't take the applied route or current URL path into account - should it? We couldn't include GET params there, since that would make the implementation vulnerable to randomisation. But if URL routing is implemented correctly, there should only be one path to get to a certain controller.

Reply:
This is blocking the merge from my perspective.

Reply:
OK - I've added a change so that the rate limit middleware accepts some extra data to be added to the key. This allows the middleware to be used across controllers/routes/etc.